xds/src/main/java/com/linecorp/armeria/xds/AbstractRoot.java

```diff
@@ -0,0 +1,130 @@
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+import io.grpc.Status;
+import io.netty.util.concurrent.EventExecutor;
+
+abstract class AbstractRoot<T extends Snapshot<? extends ResourceHolder>>
+        implements SnapshotWatcher<T>, SafeCloseable {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractRoot.class);
+    private final EventExecutor eventLoop;
+    @Nullable
+    private T snapshot;
+    private final Set<SnapshotWatcher<? super T>> snapshotWatchers = new HashSet<>();
+    private volatile boolean closed;
+
+    AbstractRoot(EventExecutor eventLoop) {
+        this.eventLoop = eventLoop;
+    }
+
+    final EventExecutor eventLoop() {
+        return eventLoop;
+    }
+
+    /**
+     * Adds a watcher which waits for a snapshot update.
+     */
+    public void addSnapshotWatcher(SnapshotWatcher<? super T> watcher) {
+        requireNonNull(watcher, "watcher");
+        checkState(!closed, "Watcher %s can't be registered since %s is already closed.",
+                   watcher, getClass().getSimpleName());
+        if (!eventLoop().inEventLoop()) {
+            eventLoop().execute(() -> addSnapshotWatcher(watcher));
+            return;
+        }
+        snapshotWatchers.add(watcher);
+        if (snapshot != null) {
+            try {
+                watcher.snapshotUpdated(snapshot);
+            } catch (Throwable t) {
+                logger.warn("Unexpected exception while invoking {}.onChanged",
+                            watcher.getClass().getSimpleName(), t);
+            }
+        }
+    }
+
+    /**
+     * Removes a watcher which waits for a snapshot update.
+     */
+    public void removeSnapshotWatcher(SnapshotWatcher<? super T> watcher) {
+        requireNonNull(watcher, "watcher");
+        checkState(!closed, "Watcher %s can't be removed since %s is already closed.",
+                   watcher, getClass().getSimpleName());
+        if (!eventLoop().inEventLoop()) {
+            eventLoop().execute(() -> removeSnapshotWatcher(watcher));
+            return;
+        }
+        snapshotWatchers.remove(watcher);
+    }
+
+    @Override
+    public void snapshotUpdated(T newSnapshot) {
+        if (closed) {
+            return;
+        }
+        assert eventLoop.inEventLoop();
+        snapshot = newSnapshot;
+        notifyWatchers("snapshotUpdated", watcher -> watcher.snapshotUpdated(snapshot));
+    }
+
+    @Override
+    public void onMissing(XdsType type, String resourceName) {
+        if (closed) {
+            return;
+        }
+        notifyWatchers("onMissing", watcher -> watcher.onMissing(type, resourceName));
```

Comment:
Every update is either propagated from

1. `SotwXdsStream` receiving an event
https://github.com/jrhee17/armeria/blob/815ab4320ae9302ad4792e878e96e20e13f03f22/xds/src/main/java/com/linecorp/armeria/xds/SotwXdsStream.java#L195

2. A timeout occurring

https://github.com/jrhee17/armeria/blob/815ab4320ae9302ad4792e878e96e20e13f03f22/xds/src/main/java/com/linecorp/armeria/xds/XdsStreamSubscriber.java#L64

So any method called should be called from an event loop.
I would rather argue that there are bigger problems if the code flow isn't called from the designated `eventLoop` since most internal data structures aren't thread-safe intentionally.

Reply:
I see. Then let's add `assert eventLoop.inEventLoop();` and `onError()` as we do in `snapshotUpdated` above.

Reply:
On second thought, I've just rescheduled from the `eventLoop` in case a user chooses to manually call `AbstractRoot#snapshotUpdated`